string_decoder : support typed array or data view

BeniCheni opened the pull request:
- [x] `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes
- [x] tests and/or benchmarks are included
- [x] documentation is changed or added
- [x] commit message follows [commit guidelines](https://github.com/nodejs/node/blob/master/doc/guides/contributing/pull-requests.md#commit-message-guidelines)

Fixes: #1826 
Per [this comment with the checklist](https://github.com/nodejs/node/issues/1826#issuecomment-410487110), this PR attempts to help out the `string_decoder` scope from the check list. The PR supports explicitly `TypedArray` or `DataView`, insead of the original `Uint8Array`.

Trott commented on lib/string_decoder.js:
```diff
@@ -72,9 +72,11 @@ StringDecoder.prototype.write = function write(buf) {
-                                   buf);
+    throw new ERR_INVALID_ARG_TYPE(
+      'buf',
+      ['Buffer', 'TypedArray', 'DataView', 'ArrayBufferView'],
```
While we're in here, does it make sense to alphabetize the list? Or is the existing ordering logical for other reasons?

addaleax replied:
I think the list makes sense in that `Buffer` will be the most common type of argument (by far), followed by `TypedArray`s.

I am not sure why `ArrayBufferView` is listed explicitly, since currently there are (afaik) only `TypedArray` and `DataView` ABVs.

Trott commented on test/parallel/test-string-decoder.js:
```diff
@@ -91,7 +91,7 @@ decoder = new StringDecoder('utf8');
 
 // A quick test for lastChar, lastNeed & lastTotal which are undocumented.
-assert(decoder.lastChar.equals(new Uint8Array([0xe1, 0, 0, 0])));
+assert(decoder.lastChar.equals(Buffer.from([0xe1, 0, 0, 0])));
```
Why this change?

Trott replied:
To elaborate a bit: There's nothing in this change that breaks `new Uint8Array()` working in this context, so why is the test case being changed instead of a new test case being added?

Trott replied:
Shouldn't test cases be added for `DataView` and `TypedArray`? Do they even work properly? No code has been added in this PR to alter how they work with `string_decoder`. Were they working already or something?

BeniCheni replied:
I'm work on the suggestion of [this comment here](https://github.com/nodejs/node/issues/1826#issuecomment-410487110):
> [ ] string_decoder
> This list is for TypedArray/DataView support only.

I was trying not use `Unit8Array`, hopefully I understood the #1826 correctly.

If this `Unit8Array` doesn't need to change to `Butter` to align with others, I can revert this change here. Please help, if I misunderstood the context of string decode in #1826.

devsnek replied:
Buffer extends Uint8Array, I don't think this changed anything.

Trott commented on lib/string_decoder.js:
```diff
@@ -72,9 +72,11 @@ StringDecoder.prototype.write = function write(buf) {
     return buf;
   if (!ArrayBuffer.isView(buf))
-    throw new ERR_INVALID_ARG_TYPE('buf',
-                                   ['Buffer', 'Uint8Array', 'ArrayBufferView'],
```
Why is `Uint8Array` being removed here?

BeniCheni replied:
[From the suggestion of this comment in #1826](https://github.com/nodejs/node/issues/1826#issuecomment-410487110), "string_decoder" seems to be on the list `TypedArray`/`DataView` support only. After viewing the [fs PR](https://github.com/nodejs/node/pull/22150) and the [child_process](https://github.com/nodejs/node/pull/22409), I just gave my best attempt to follow the "_idea_" to handle string_decode.

Please ;et me know if I misunderstood too much off from the context of the checklint & goals from the comment.

addaleax replied:
I think this is okay and somewhat consistent with how we handle it elsewhere

SirR4T replied:
Hi @BeniCheni , I think broadly we can categorize the changes for these PRs into:
1. adding missing tests, for `TypedArray`s and `DataView`s wherever only `Buffer`s or `Uint8Array`s were tested earlier
2. Updating code to work with these new types, where necessary(should mostly involve only JS side of things, in `lib` and maybe `lib/internal`)
3. Updating documentation to reflect these changes.

Here, your PR is updating the errors being thrown, but is neither adding test cases nor adding functionality to work with the `TypedArray`s or `DataView`s. Please note, it might well be the case that `string_decoder` already works for these, but adding test cases is a way of validation for the same.  Hope that helps!

BeniCheni replied:
🙏 for the collective reviews. Will analyze & update accordingly.

BeniCheni commented:
> Shouldn't test cases be added for DataView and TypedArray?

Researched & write tests to cover `TypedArray` & `DataView`.

> Why is Uint8Array being removed here?
> Buffer extends Uint8Array, I don't think this changed anything.

Researched more and updated to revert the previously incorrect removing `Uint8Array`.

> LGTM when ArrayBufferView is removed.

Removed any context of `ArrayBufferView` in `string_decoder` source per suggestion.

---

Also confirmed if any context needed to update in `string_decoder.md` doc, but didn't find any necessary edit. Please suggest if I miss anything in the `string_decoder.md` doc. Please review again.

SirR4T commented:
Hi @BeniCheni , thanks for updating the tests! I think we could also update `string_decoder.md`, specifically for [the `.end()`](https://nodejs.org/api/string_decoder.html#string_decoder_stringdecoder_end_buffer) and [the `.write()`](https://nodejs.org/api/string_decoder.html#string_decoder_stringdecoder_write_buffer) function documentations, since they accept not just buffers, but other types too.

SirR4T commented:
Hi @BeniCheni , doc changes look good to me, thanks!

But as I'm myself a new enough contributor here, I'll refrain from approving PRs, in deference to someone with more review experience. 

As for the CI failure, it doesn't seem related to your changes, but instead looks like a certificate failure from npm. This may or may not be related to [npm's current issue](https://status.npmjs.org/incidents/v22ffls5cd6h) at the time of this post.

addaleax commented:
New CI: https://ci.nodejs.org/job/node-test-pull-request/16953/

addaleax commented:
Resume CI: https://ci.nodejs.org/job/node-test-pull-request/16971/

TimothyGu commented on test/parallel/test-string-decoder.js:
```diff
@@ -97,6 +97,21 @@ assert.strictEqual(decoder.lastTotal, 3);
 
+// TypedArray with Int8Array test
+decoder = new StringDecoder('utf8');
+assert.strictEqual(decoder.write(new Int8Array([1, 2])), '\u0001\u0002');
```
I'd prefer this test use [getArrayBufferViews()](https://github.com/nodejs/node/blob/master/test/common/README.md#getarraybufferviewsbuf) to cover all possible TypedArray types and DataView. An example usage of this function by @SirR4T is

https://github.com/nodejs/node/blob/be5e3964b316da013d02ee743420e5fd9cb50c76/test/parallel/test-fs-write-file-typedarrays.js#L24-L31

BeniCheni commented:
Dear reviewer(s),
Friendly note to check if there’s anything pending for this “author ready” issue? Thanks~

jasnell commented on test/parallel/test-string-decoder.js:
```diff
@@ -97,6 +97,21 @@ assert.strictEqual(decoder.lastTotal, 3);
+const arrayBufferViewStr = 'String for ArrayBufferView tests\n';
+const inputBuffer = Buffer.from(arrayBufferViewStr.repeat(8), 'utf8');
+for (const expectView of common.getArrayBufferViews(inputBuffer)) {
+  console.log(
```
If the `console.log()` statement is part of the test, please add a comment indicating such. If it is not, I'd prefer to use a comment instead of having the extraneous console output.

BeniCheni commented:
> @ jasnell 
> If the console.log() statement is part of the test, please add a comment indicating such. If it is not, I'd prefer to use a comment instead of having the extraneous console output.

Hello @jasnell,
I've checked & removed the unneeded console log message from the string_decode test code. Please review at convenience. Thanks.

BeniCheni commented:
Ping/Dear reviewers,

Would you mind helping to guide this “author ready” PR to landing?

(Made all suggested updates according to previous review comments; thanks!)

addaleax commented:
New CI: https://ci.nodejs.org/job/node-test-pull-request/17187/

addaleax commented:
Landed in e2325bcc0415a86a77faf06a8e85503db9f6618f
